eos-esp-generator

```diff
@@ -0,0 +1,968 @@
+    return data
+
+
+def get_partition_data():
+    """Gather disk partition data
+
+    Returns a list of partition dicts. Ideally this would use lsblk --json, but
+    that just collects udev attributes. Since this will be run as a generator,
+    udev won't be running yet. In that case, blkid is used to gather partition
+    data just like udev itself does.
+    """
+    partitions = []
+
+    # Get the partition devices, stripping any empty lines in the output.
+    proc = run(
+        ['blkid', '-d', '-c', '/dev/null', '-o', 'device'],
+        stdout=subprocess.PIPE,
+    )
+    partition_devs = [line for line in proc.stdout.splitlines() if line]
+    logger.debug(f'partition devices: {partition_devs}')
+
+    # Probe all the partition devices.
+    for dev in partition_devs:
+        try:
+            proc = run(
+                ['blkid', '--probe', '-d', '-c', '/dev/null', '-o', 'export', dev],
+                stdout=subprocess.PIPE,
+            )
+        except subprocess.CalledProcessError as err:
+            # blkid will exit with code 2 if it can't gather information about
+            # the device. Just carry on in that case.
+            if err.returncode == 2:
+                logger.warning(f'Ignoring blkid --probe exit code 2 for {dev}')
+                continue
+            else:
+                raise
+
+        entry = {}
+        for line in proc.stdout.splitlines():
+            tag, _, value = line.partition('=')
+            if not tag:
+                continue
+            entry[tag] = value
+
+        if entry:
+            partitions.append(entry)
+        else:
+            logger.warning(f'No blkid probe output for {dev}')
+
+    logger.debug('partitions:')
+    for entry in partitions:
+        logger.debug(f' {entry}')
+    return partitions
+
+
+def read_efivar(name):
+    """Read an EFI variable value
+
+    Reads the data from efivarfs mounted at /sys/firmware/efi/efivars.
+    Only the variable value is returned, not the attributes.
+    """
+    root = _get_root_path()
+    varpath = root / 'sys/firmware/efi/efivars' / name
+    logger.debug(f'Reading EFI variable {varpath}')
+    try:
+        with open(varpath, 'rb') as f:
+            value = f.read()
+    except FileNotFoundError:
+        logger.debug(f'EFI variable {name} not set')
+        return None
+
+    # Skip the first 4 bytes, those are the 32 bit attribute mask.
+    if len(value) < 4:
+        logger.warning(f'Invalid EFI variable {name} is less than 4 bytes')
+        return None
+    return value[4:]
+
+
+def read_efivar_utf16_string(name):
+    """Read an EFI variable UTF-16 string
+
+    If the EFI variable doesn't exist, None is returned. Any nul
+    terminating bytes will be removed.
+    """
+    value = read_efivar(name)
+    if value is None:
+        return None
+
+    logger.debug(f'EFI variable {name} contents: {value}')
+
+    # Systemd appends 3 nul bytes for some reason. If there are an odd
```

Comment:
I'm curious to know what the reason is!

(From a less careful codebase I might wonder if this is the UTF-16-encoding of U+0000 (two nul bytes) followed by a C byte string terminator (1 nul byte).)

Reply:
I found the spot in the systemd code where it happens, but it was quite a while ago. There weren't really comments, but I think it was sort of what you're suggesting. Whether you try to parse as a C string or UTF-16, you'll find a trailing NUL. But it just has easily could have been a mistake that's essentially irrelevant the way systemd reads the values. The Unicode handling was entirely bespoke. I'll try to find it again when I get a chance.